src/richtext/utils/hastUtilToRichText.ts

```diff
@@ -0,0 +1,373 @@
+	 */
+	context: {
+		/**
+		 * The list type of the last list node encountered if any.
+		 */
+		listType: "group-list-item" | "group-o-list-item" | null;
+	};
+};
+
+/**
+ * Serializes a hast {@link Element} node to a
+ * {@link RichTextHTMLMapSerializerShorthand} or a rich text node.
+ *
+ * @remarks
+ * Serializing to a rich text node directly is not recommended and is only
+ * available as an escape hatch. Prefer returning a
+ * {@link RichTextHTMLMapSerializerShorthand} instead.
+ */
+type RichTextHTMLMapSerializerFunction = (
+	payload: RichTextHTMLMapSerializerFunctionPayload,
+) =>
+	| RichTextHTMLMapSerializerShorthand
+	| RTNode
+	| RTInlineNode
+	| RTPartialInlineNode
+	| null
+	| undefined;
+
+/**
+ * Serializes an hast {@link Element} node matching the given HTML tag name or
+ * CSS selector to a Prismic rich text node.
+ */
+export type RichTextHTMLMapSerializer = Record<
+	string,
+	RichTextHTMLMapSerializerShorthand | RichTextHTMLMapSerializerFunction
+>;
+
+const DEFAULT_SERIALIZER: RichTextHTMLMapSerializer = {
+	h1: "heading1",
+	h2: "heading2",
+	h3: "heading3",
+	h4: "heading4",
+	h5: "heading5",
+	h6: "heading6",
+	p: "paragraph",
+	pre: "preformatted",
+	strong: "strong",
+	em: "em",
+	li: ({ context }) =>
+		context.listType === "group-o-list-item" ? "o-list-item" : "list-item",
+	ul: "group-list-item",
+	ol: "group-o-list-item",
+	img: "image",
+	iframe: "embed",
+	a: "hyperlink",
+};
+
+const VFILE_RULE = "failed-to-serialize-node";
+const VFILE_SOURCE = "prismic";
+
+/**
+ * Configuration that determines the output of `hastUtilToRichText`.
+ */
+export type HastUtilToRichTextConfig = {
+	/**
+	 * An optional HTML to rich text serializer. Will be merged with the default
+	 * HTML to rich text serializer.
+	 */
+	serializer?: RichTextHTMLMapSerializer;
+
+	/**
+	 * Whether or not the text processed should be marked as right-to-left.
+	 *
+	 * @defaultValue `false`
+	 */
+	direction?: "ltr" | "rtl";
+
+	/**
+	 * The default node type to wrap the input with in case it only represents the
+	 * inner content of an element.
+	 *
+	 * @remarks
+	 * This is a niche option that allows you to serialize correctly an input such
+	 * as: `"lorem <strong>ipsum</strong> dolor sit amet"`, wrapping it inside a
+	 * node of the desired type.
+	 *
+	 * @defaultValue `"paragraph"`
+	 */
+	defaultWrapperNodeType?: RTTextNode["type"];
+};
+
+export const hastUtilToRichText = (
+	tree: Root | Element,
+	file: VFile,
+	config?: HastUtilToRichTextConfig,
+): RichTextField => {
+	const builder = new RichTextFieldBuilder();
+
+	// Merge the default serializer with the user-provided one.
+	const serializer = {
+		...DEFAULT_SERIALIZER,
+		...config?.serializer,
+	};
+
+	// Keep track of the last text node type to append text nodes to in
+	// case of an image or an embed node is present inside a paragraph.
+	let lastRTTextNodeType: RTTextNode["type"] =
+		config?.defaultWrapperNodeType ?? RichTextNodeType.paragraph;
```

Comment:
I believe they can since images are inline tags, but since we can't ensure a `heading1` node to be necessarily represented by an `h1` tag (it could be a `div` or anything), technically this makes it valid HTML, so this is all an escape hatch for such cases.

Reply:
I believe they can since images are inline tags, but since we can't ensure a `heading1` node to be necessarily represented by an `h1` tag (it could be a `div` or anything), technically this makes it valid HTML, so this is all an escape hatch for such cases.

Reply:
I'm not clear on the use case of `defaultWrapperNodeType`. I see how it affects the output and that it's really only for cases where you don't have a preceding text node, but I'm not sure where a user would want something besides a paragraph.

Do you have an idea?

Reply:
I'm not strongly against the option, but I'm wondering if defaulting to `paragraph` is sufficient.

Reply:
I introduced it after seeing some WordPress ACF fields could only return the inner HTML content of a node. (Example was for an author bio here on the get method example: https://www.advancedcustomfields.com/resources/wp-rest-api-integration/)

I agree however I don't see what else of a paragraph you'd want here, so maybe we can remove it and default to a paragraph node, and later reintroduce it if it's actually needed.

Reply:
Removed for now!